### PR TITLE
Use mod() vs fmod() logic for OSL and MDL codegen

### PR DESCRIPTION
**Pull request by bernardkwok**
Update #870 
# OSL
- Change fmod() usage to mod() usage for OSL implementations of modulo.
- Added reference implementations for vector2,vector4,color2,color4 and used for inlines
# MDL
 - Add in mx_mod() implementation to core + use for implementations currently using math::fmod().
- Add MDL module install path option for release and testing
- Fix up test compile code to use new module path option
# Test Results:
[genmdl_mdl_generate_test.txt](https://github.com/autodesk-forks/MaterialX/files/4600940/genmdl_mdl_generate_test.txt)

### CODE REVIEW

**Comment by bernardkwok** [comment]
@niklasharrysson, review if you have time. Otherwise I've added @ashwinbhat to review. Thanks.

File: CMakeLists.txt [hunk 1]
```diff
@@ -91,9 +91,12 @@ set(MATERIALX_TESTRENDER_EXECUTABLE "" CACHE FILEPATH "Full path to the testrend
...
 
 # Helpers for MDL validation
-set(MATERIALX_MDLC_EXECUTABLE "" CACHE FILEPATH "Full path to the mdlc binary.")
-set(MATERIALX_MDL_RENDER_EXECUTABLE "" CACHE FILEPATH "Full path to the mdl renderer binary.")
-set(MATERIALX_MDL_MODULE_PATHS "" CACHE FILEPATH "Comma separated list of MDL module paths.")
+if (MATERIALX_BUILD_GEN_MDL)
+    set(MATERIALX_MDLC_EXECUTABLE "" CACHE FILEPATH "Full path to the mdlc binary.")
+    set(MATERIALX_MDL_RENDER_EXECUTABLE "" CACHE FILEPATH "Full path to the mdl renderer binary.")
+    set(MATERIALX_MDL_MODULE_PATHS "" CACHE FILEPATH "Comma separated list of MDL module paths.")
+    set(MATERIALX_INSTALL_MDL_MODULE_PATH ${MATERIALX_INSTALL_STDLIB_PATH} CACHE FILEPATH "Install path for mdl module.")
```
Comment (bernardkwok) [inline CMakeLists.txt:98]: Add new MDL module path for installation and test.

File: libraries/stdlib/genmdl/stdlib_genmdl_impl.mtlx [hunk 2]
```diff
@@ -285,19 +285,19 @@
   <implementation name="IM_divide_matrix44_genmdl" nodedef="ND_divide_matrix44" sourcecode="{{in1}}" language="genmdl" /> <!-- TODO: Implement properly -->
 
   <!-- <modulo> -->
-  <implementation name="IM_modulo_float_genmdl" nodedef="ND_modulo_float" sourcecode="math::fmod({{in1}}, {{in2}})" language="genmdl"/>
-  <implementation name="IM_modulo_color2_genmdl" nodedef="ND_modulo_color2" sourcecode="math::fmod({{in1}}, {{in2}})"  language="genmdl"/>
-  <implementation name="IM_modulo_color2FA_genmdl" nodedef="ND_modulo_color2FA" sourcecode="math::fmod({{in1}}, {{in2}})" language="genmdl"/>
+  <implementation name="IM_modulo_float_genmdl" nodedef="ND_modulo_float" sourcecode="mx_mod({{in1}}, {{in2}})" language="genmdl"/>
```
Comment (bernardkwok) [inline libraries/stdlib/genmdl/stdlib_genmdl_impl.mtlx:288]: Change to use new functions vs built-in fmod.

File: libraries/stdlib/osl/color2.h [hunk 3]
```diff
@@ -242,6 +242,22 @@ color2 min(color2 a, float b)
                   min(a.a, b));
 }
 
+color2 mod(color2 a, color2 b)
```
Comment (bernardkwok) [inline libraries/stdlib/osl/color2.h:245]: Patches in a few places to have function overloads for mod() variants of fmod()

File: source/MaterialXGenMdl/mdl/materialx/core.mdl [hunk 4]
```diff
@@ -99,6 +99,13 @@ export color4 mx_multiply( color4 v1, color4 v2) { return color4(v1.rgb * v2.rgb
 export color4 mx_multiply( color4 v1,  float v2) { return color4(v1.rgb * v2    , v1.a * v2  ); }
 export color4 mx_divide(   color4 v1, color4 v2) { return color4(v1.rgb / v2.rgb, v1.a / v2.a); }
 export color4 mx_divide(   color4 v1,  float v2) { return color4(v1.rgb / v2    , v1.a / v2  ); }
+export float mx_mod(       float x, float y) { return x - y * math::floor(x/y); }
```
Comment (bernardkwok) [inline source/MaterialXGenMdl/mdl/materialx/core.mdl:102]: New function overloads for mod.

File: source/MaterialXTest/MaterialXGenMdl/GenMdl.cpp [hunk 5]
```diff
@@ -132,8 +132,9 @@ void MdlShaderGeneratorTester::compileSource(const std::vector<mx::FilePath>& so
     {
         mdlcCommand += " -p\"" + extraPath + "\"";
     }
+    // Note: These paths are based on
     mx::FilePath currentPath = mx::FilePath::getCurrentPath();
-    mx::FilePath coreModulePath = currentPath / mx::FilePath("libraries/stdlib/genmdl");
+    mx::FilePath coreModulePath = currentPath / std::string(MATERIALX_INSTALL_MDL_MODULE_PATH) / "mdl");
```
Comment (bernardkwok) [inline source/MaterialXTest/MaterialXGenMdl/GenMdl.cpp:137]: Make MDL gen test use correct module path.

File: source/MaterialXTest/MaterialXGenMdl/GenMdl.cpp [hunk 6]
```diff
@@ -177,7 +178,7 @@ TEST_CASE("GenShader: MDL Shader Generation", "[genmdl]")
 
     const mx::FilePath logPath("genmdl_mdl_generate_test.txt");
 
-    bool writeShadersToDisk = false;
+    bool writeShadersToDisk = true;
```
Comment (bernardkwok) [inline source/MaterialXTest/MaterialXGenMdl/GenMdl.cpp]: This will produce code but not run the mdlc test unless specified (which is false by default).

**Comment by ashwinbhat** [comment]
Adding @nicolassavva-autodesk. Not sure if this impacts adsklib

**Comment by bernardkwok** [comment]
@nicolassavva-autodesk :
I looked over adsklib and see this:

` specular_rotation = fmod(surface_rotation / 360.0, 1.0)`

For this I assume surface_rotation should never be < 0 in which case fmod / mod are the same.
BTW, this just makes modulo for MDL and OSL consistent with GLSL (though AFAIK GLSL mod is not the same as HLSL fmod).

**Comment by ashwinbhat** [comment]
> @nicolassavva-autodesk :
> I looked over adsklib and see this:
> 
> ` specular_rotation = fmod(surface_rotation / 360.0, 1.0)`
> 
> For this I assume surface_rotation should never be < 0 in which case fmod / mod are the same.
> BTW, this just makes modulo for MDL and OSL consistent with GLSL (though AFAIK GLSL mod is not the same as HLSL fmod).

Sounds good.
Yeah GLSL/HLSL is different: https://forum.unity.com/threads/translating-a-glsl-shader-noise-algorithm-to-hlsl-cg.485750/

**Pull request merged**